woocommerce/woocommerce-ios
# HACK week: Add coupon to Order [2/2]

Part 2/2 of HACK week project: Add coupon to Order
- [x] ⚠️  Make sure to review and merge part 1 #9089
- [x] ⚠️  Log GH issue to track pending work before merging this PR @selanthiraiyan https://github.com/woocommerce/woocommerce-ios/issues/9111

## Description

This PR adds an option to add a coupon to order during create/update flows

## Testing instructions

- Launch the app and login if needed
- Ensure that your site has available coupons

**Create new order**
- Navigate to `Orders` tab and tap on tap on `+`
- Tap on `+ Add Product` and add a product
- Tap on `+ Add Coupon` to open coupon screen
- Enter a valid coupon code and tap `Done` 
- Verify that you can successfully create an Order with a coupon code 
- Ensure that you see the following analytics events in Xcode console
```
Tracked order_coupon_add, properties: [AnyHashable("flow"): "creation"]
```

**Update existing order**
- Navigate to `Orders` tab and open an existing order with coupon
- Tap `Edit` on the top right to open the edit order screen
- Tap on `Coupon` row to edit the coupon
- In the textfield remove the text and enter a valid coupon code
- Tap `Done` and verify that you can successfully update a coupon to an order

**Remove coupon**
- Navigate to `Orders` tab and open an existing order with coupon
- Tap `Edit` on the top right to open the edit order screen
- Tap on `Coupon` row to edit the coupon
- Tap `Remove Coupon from Order` to delete the coupon
- Verify that the coupon is removed successfully from the order
- Ensure that you see the following analytics events in Xcode console
```
Tracked order_coupon_remove, properties: [AnyHashable("flow"): "editing"]
```

**Test feature flag**
- Disable the `addCouponToOrder` feature flag by returning `false` [here](https://github.com/woocommerce/woocommerce-ios/pull/9089/files#diff-26c9afc49bb1f1a168517d43d5e36a64ff45adb046469369ef3e4d6800ccc19bR87)
- Navigate to `Orders` tab and tap on tap on `+`
- Ensure that you can't see `+ Add a coupon` row
- Go back and navigate to the `Orders` tab and open an existing order with coupon
- Tap `Edit` on the top right to open the edit order screen
- Ensure that you can't see the `Coupon` row 

### Known issue
- We need to show a loading indicator and disable further editing when a coupon is deleted
- To know when a coupon has been deleted, we use this code [now](https://github.com/woocommerce/woocommerce-ios/pull/9090/files#diff-0667e76ef83e3a4ed8c121e825fa2a1fcc649011939b3755888c8216d1fa8c0eR450-R456)
- I reached out to @hichamboushaba to validate this, and he helpfully pointed out that it is not safe to assume that an order’s `discountTotal` will match the total of coupon line’s `discount`s. Internal slack here - p1678280360582379/1678264651.362909-slack-C03L1NF1EA3
- Since this feature is behind a flag, I will work on a fix before we release this. 
- I will log a GH issue before I merge this PR.

## Screenshots
https://user-images.githubusercontent.com/524475/223976463-c59f777d-f18e-4190-8c96-ea1dd3559bfa.mp4

---
